hibernate-core/src/main/java/org/hibernate/query/sqm/internal/SqmCriteriaNodeBuilder.java

```diff
@@ -1451,6 +1451,25 @@ private static TrimSpec convertTrimSpec(Trimspec jpaTs) {
 		}
 	}
 
+	@SuppressWarnings("unchecked")
+	private <T> BasicType<T> guessType(T value) {
+		BasicType<T> basicType;
+		Class<T> valueClass = (Class<T>) value.getClass();
+		basicType = getTypeConfiguration().getBasicTypeForJavaType( valueClass );
+		if ( basicType == null ) {
+			final JavaType<Object> javaType = getTypeConfiguration().getJavaTypeDescriptorRegistry()
+					.getDescriptor( valueClass );
+			if ( javaType != null ) {
+				final JdbcType recommendedJdbcType = javaType.getRecommendedJdbcType( getTypeConfiguration().getCurrentBaseSqlTypeIndicators() );
+				if ( recommendedJdbcType != null ) {
+					basicType = (BasicType<T>) getTypeConfiguration().getBasicTypeRegistry().resolve( javaType, recommendedJdbcType);
+				}
+			}
```

Comment:
I think that's not totally correct as you can see from the test failures for e.g. PostgreSQL, because `JavaTypeRegistry#getDescriptor` creates a `SerializableJavaTypeDescriptor` or `JavaTypeDescriptorBasicAdaptor` on-demand if there it can't find an existing JavaType for the class.
IMO we should introduce a `JavaTypeRegistry#findDescriptor` that returns null if the descriptor does not exist rather than trying to create a new one.

Reply:
I merged my PR for this, you can rebase and then use `findDescriptor` here instead: https://github.com/hibernate/hibernate-orm/pull/4359

Reply:
Could you have a look at the changes Steve did to in commit 0b4696609 to AbstractJdbcParameter? Since I based my code on his there, maybe it needs changing also?